zebra-chain/src/amount.rs

```diff
@@ -191,6 +191,67 @@ where
     }
 }
 
+impl std::ops::Mul<u64> for Amount<NonNegative> {
+    type Output = Result<Amount<NonNegative>>;
+
+    fn mul(self, rhs: u64) -> Self::Output {
+        let value = (self.0 as u64)
+            .checked_mul(rhs)
+            .ok_or(Error::MultiplicationOverflow {
+                amount: self.0,
+                multiplier: rhs,
+            })?;
+        value.try_into()
+    }
+}
+
+impl std::ops::Mul<Amount<NonNegative>> for u64 {
+    type Output = Result<Amount<NonNegative>>;
+
+    fn mul(self, rhs: Amount<NonNegative>) -> Self::Output {
+        rhs.mul(self)
+    }
+}
+
+impl std::ops::MulAssign<u64> for Amount<NonNegative> {
+    fn mul_assign(&mut self, rhs: u64) {
+        // overflow will return `Amount<NonNegative>` unchanged
+        self.0 = (self.0 as u64).checked_mul(rhs).unwrap_or(self.0 as u64) as i64
+    }
+}
+
+impl std::ops::Div<u64> for Amount<NonNegative> {
+    type Output = Result<Amount<NonNegative>>;
+
+    fn div(self, rhs: u64) -> Self::Output {
+        let quotient = (self.0 as u64)
+            .checked_div(rhs)
+            .ok_or(Error::DivideByZero { amount: self.0 })?;
+        // since this is a division by a positive integer,
+        // the quotient must be within the constrained range
+        Ok(quotient.try_into().unwrap())
+    }
+}
+
+impl std::ops::DivAssign<u64> for Amount<NonNegative> {
+    fn div_assign(&mut self, rhs: u64) {
+        // division by zero will return `Amount<NonNegative>` unchanged
+        self.0 = (self.0 as u64).checked_div(rhs).unwrap_or(self.0 as u64) as i64
+    }
+}
+
+impl PartialOrd for Amount<NonNegative> {
+    fn partial_cmp(&self, other: &Amount<NonNegative>) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for Amount<NonNegative> {
+    fn cmp(&self, other: &Amount<NonNegative>) -> std::cmp::Ordering {
+        self.0.cmp(&other.0)
+    }
+}
```

Comment:
@yaahc do you think we should make `Ord for Amount` generic over `NonNegative` and `NegativeAllowed`?

If we make it generic, let's also move `Ord` and `PartialOrd` to:
https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/amount.rs

Reply:
The use case for comparing NonNegative with NonNegative is this consensus rule:
> the sum of the coinbase transaction outputs must be less than or equal to the block subsidy plus transaction fees

Do you think we'll want to compare NegativeAllowed with NegativeAllowed, or NonNegative with NegativeAllowed?

Reply:
I think it will be needed, added athttps://github.com/ZcashFoundation/zebra/pull/1051/commits/be5645b8981433c7b35486d79c8dc6c96a8bf633